Trait formatting

dingo-d opened the pull request:
This PR depends on #100. Once that is merged, this PR should be rebased to the master branch.

The PR adds rules about the trait formatting rules and is the continuation of the additions of 'modern' PHP code in the WordPress PHP Coding Standards handbook based on the [make post](https://make.wordpress.org/core/2020/03/20/updating-the-coding-standards-for-modern-php/) by Juliette Reinders Folmer.

jrfnl commented on wordpress-coding-standards/php.md:
````diff
@@ -366,6 +366,49 @@ $a = foo(
 );
 ```
 
+### Trait Formatting
````
```suggestion
### Trait Use statements
```

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -366,6 +366,49 @@ $a = foo(
+    ...
+}
+
+class Foo {
+
+    use Bar_Trait;
+}
```
I think this code sample can be omitted.

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -366,6 +366,49 @@ $a = foo(
+// Correct.
+class Foo {
+
+    // Comments above a use statement will be regarded as documentation for the statement.
```
I think this comment line can be omitted.

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -366,6 +366,49 @@ $a = foo(
+    use       Bar_Trait; // No blank line before class declaration, multiple spaces after the use statement.
+
+
+    use Foo_Trait,   Bazinga_Trait{Bar_Trait::method_name    insteadof     Foo_Trait; Bazinga_Trait::method_name // Multiple spaces when importing traits, no new line after opening brace
+        as     bazinga_method; // Multiple spaces, should be on the same line as the method it's replacing.
+            }; // Wrongly indented brace.
+    public $baz = true; // Missing blank line after trait import.
```
This code is hard to read as it is (which is exactly the point) and the comments make it even harder.

Maybe list out the things which are wrong in a multi-line `/* .. */` comment above the `use` statement ?

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -366,6 +366,49 @@ $a = foo(
+
+Trait `use` statements should be at the top of a class and should have exactly one blank line before the first `use` statement, and at least one blank line after. The only exception is when the class only contains the trait `use` statement, in which case the blank line after may be omitted.
+
+The following code examples cover the rule about formatting traits regarding things like spacing, grouping and indentation.
```
```suggestion
The following code examples show the formatting requirements for trait `use` statements regarding things like spacing, grouping and indentation.
```

jrfnl reviewed (commented):
@dingo-d Thanks again! Reviewed.

---

I wonder if this section is placed correctly.

PSR12 has these kind of rules in the ["Classes, Properties, and Methods" chapter](https://www.php-fig.org/psr/psr-12/#42-using-traits) below the "Declare Statements, Namespace, and Import Statements" chapter.

In the WPCS handbook page, that would translate to moving this section to the "Object-Oriented Programming" chapter, which sounds about right.

What do you think ?

---

When comparing the section to the Make post, I'm missing some things.
* Under "Naming Conventions", line 91 "Class names should use capitalized words separated by underscores. Any acronyms should be all upper case." should probably be updated to include traits, interfaces and enums.
    Might be a nice idea to also add an example of one of these constructs to the code sample just below it (or replace one of the existing code samples).
* ~~Under "Naming Conventions", line 110 "Class file names should be based on the class name with `class-` prepended and the underscores in the class name replaced with hyphens" should probably also be updated to include traits, interfaces and enums.~~ **_Actually, you were right to not include this one as there was some unresolved discussion about that part of the proposal_**.
* "For everything else, the same formatting rules as already in place for class declarations apply." - to be honest aside from line 91 and the code sample below it, I can't really find any rules in the handbook about class formatting... oops. Let's revisit that later...

---

Also note: there is still a merge conflict and a duplicate section in the current commits (both for the fully qualified names section).

dingo-d commented:
> In the WPCS handbook page, that would translate to moving this section to the "Object-Oriented Programming" chapter, which sounds about right.

I thought about this, but because we've had _Trait and Interface naming_ as one chapter (which is TBD) and _Trait and interface formatting_ (which was in fact just Trait formatting), I've changed the name to just Trait formatting and move it to the Formatting section.

If we were to rename the chapter to the suggested _Trait Use Statements_ then it would fit in the OOP section.

Do you want to go in that direction?

> Merge conflicts

I guess this is the side effect of rebasing I did 🙈 I'll apply your commits and rebase to the master branch once more and fix the merge conflicts.

dingo-d commented:
@jrfnl I've moved the traits use section to the OOP section, removed duplicate sections, and added the suggestions from the PR review.

Hope it looks better now.

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -507,6 +511,49 @@ class Example_Class { [...] }
+// Incorrect.
+class Foo {
+    /**
+     * No blank line before class declaration, multiple spaces after the use statement.
```
+ end of sentence: "use statement" => "`use` keyword"

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -507,6 +511,49 @@ class Example_Class { [...] }
+
+// Incorrect.
+class Foo {
+    /**
```
I concur. Docblock is the wrong format. Either multiple `//` or a multi-line inline comment `/* ... */` (note: single `*`).

jrfnl commented on wordpress-coding-standards/php.md:
```diff
@@ -507,6 +511,49 @@ class Example_Class { [...] }
 
+### Trait Use Statements
+
+Trait `use` statements should be at the top of a class and should have exactly one blank line before the first `use` statement, and at least one blank line after. The only exception is when the class only contains the trait `use` statement, in which case the blank line after may be omitted.
```
> Going from potentially an infinite amount of blank lines after, to potentially zero would seem to make it harder to enforce than necessary. Can we change the "at least one blank line after" to [exactly] "one blank line after"? That way, it's either one line after, or zero lines after if that's the only thing in the class.

This was a very deliberate choice of wording, both for the class only containing trait `use` as well as the class with more content. We use the same wording in multiple other places too.

In this case, whatever _follows_ the trait `use` statement determines how many blank lines _between_ the `use` and the next thing are allowed. The trait `use` "blank lines after" only determines that there should be _at least_ one blank line.

In more general terms: strict rules for blank lines before something, loose rules for blank lines after.

If we don't do it this way, it can very easily lead to fixer conflicts, which is something we want to avoid.

As for the "class only containing trait `use` may omit the blank line after" part, again, this is largely to prevent fixer conflicts as generally speaking, there should be no stray blank lines at the end of a class and sniffs checking for that normally don't care what is directly above the closing brace, as long as it isn't a blank line.

jrfnl replied:
@GaryJones Does it make more sense when you read it with that context in mind ?

jrfnl replied:
@dingo-d Looks like @GaryJones' suggestion to remove ambiguity hasn't been applied yet.

dingo-d commented:
I've fixed the issues and we can merge the PR. We can always discuss the parts that @GaryJones had issues within a subsequent issue/PR (Juliette gave good reasoning behind choosing the wording in that paragraph).

jrfnl reviewed (approved):
👍🏻 

Merging this now as @GaryJones is AFK for the next few weeks. If needs be, additional textual tweaks can still be made in a follow up commit.
